### PR TITLE
Phase 4 controller merging

### PR DESCRIPTION
**Pull request by Kegsay**
This PR is the final phase to finish merging all controller/views between react SDK and Vector.

The following additional changes have been made:
 - Move `src/DateUtils.js`and `src/Resend.js` from vector to react SDK.
 - Dependency inject a `ConferenceHandler` into `MatrixChat` in vector-web's `src/vector/index.js`. This removes the need for the `Modulator`.
 - Move the `Notifier` to be a pure logic class like everything else. It was masquerading as a UI component even though it's not a React class.

After this PR lands we're mostly over the hill on the restructure. See https://github.com/vector-im/vector-web/pull/448 for the vector changes.

### CODE REVIEW

File: src/Notifier.js [hunk 1]
```diff
@@ -16,8 +16,10 @@ limitations under the License.
```
Comment (dbkr) [inline src/Notifier.js]: Is there a reason for changing this into a non-replaceable component? Surely lots of apps will want to present their notifications differently?

File: src/components/structures/MatrixChat.js [hunk 2]
```diff
@@ -468,5 +487,181 @@ module.exports = {
...
+            return (
+                <div className="mx_MatrixChat_splash">
+                    <Spinner />
+                    <a href="#" className="mx_MatrixChat_splashButtons" onClick={ this.onLogoutClick }>Logout</a>
+                </div>
+            );
+        } else if (this.state.screen == 'register') {
+            return (
+                <Registration
+                    clientSecret={this.state.register_client_secret}
```
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: I don't really get why it makes sense for MatrixChat to have to know about details of the registration process like the client secret.

File: src/components/structures/login/Login.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        homeserverUrl: React.PropTypes.string,
+        identityServerUrl: React.PropTypes.string,
+        // login shouldn't know or care how registration is done.
+        onRegisterClick: React.PropTypes.func.isRequired
+    },
+
+    getDefaultProps: function() {
+        return {
+            homeserverUrl: 'https://matrix.org/',
+            identityServerUrl: 'https://vector.im'
```
Comment (dbkr) [inline src/components/structures/login/Login.js]: react-sdk now seems to be defaulting to vector.im. What happened to the bit that reads these from the config file?

File: src/components/structures/login/Login.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                    <CasLogin />
+                );
+            default:
+                if (!step) {
+                    return;
+                }
+                return (
+                    <div>
+                    Sorry, this homeserver is using a login which is not
+                    recognised by Vector ({step})
```
Comment (dbkr) [inline src/components/structures/login/Login.js]: Vector branding

File: src/components/structures/login/Login.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                                { this.state.errorText }
+                        </div>
+                        <a className="mx_Login_create" onClick={this.props.onRegisterClick} href="#">
+                            Create a new account
+                        </a>
+                        <br/>
+                        <div className="mx_Login_links">
+                            <a href="https://medium.com/@Vector">blog</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
+                            <a href="https://twitter.com/@VectorCo">twitter</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
+                            <a href="https://github.com/vector-im/vector-web">github</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
```
Comment (dbkr) [inline src/components/structures/login/Login.js]: More vectors

File: src/components/views/login/ServerConfig.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+    showHelpPopup: function() {
+        var ErrorDialog = sdk.getComponent('dialogs.ErrorDialog');
+        Modal.createDialog(ErrorDialog, {
+            title: 'Custom Server Options',
+            description: <span>
+                You can use the custom server options to log into other Matrix
+                servers by specifying a different Home server URL.
+                <br/>
+                This allows you to use Vector with an existing Matrix account on
```
Comment (dbkr) [inline src/components/views/login/ServerConfig.js]: Vectorism

**Comment by dbkr** [comment]
Difficult to tell what's changed and what's just been move din from vector, but I think otherwise this is looking okay.

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (Kegsay) [inline src/components/structures/MatrixChat.js]: Completely agree, but things like `clientSecret` are obtained from the URL as query parameters. They need to be passed through somehow.

File: src/components/structures/login/Login.js (hunk 3 above)
Comment (Kegsay) [inline src/components/structures/login/Login.js]: Still included, see https://github.com/vector-im/vector-web/pull/448/files#diff-fc81839358945a06917a3c653b7a76ceR149 - It should be `matrix.org` here though, so will fix.

**Comment by Kegsay** [comment]
WRT Vector branding, I'm aware they need to be `matrix.org`-ified but have chosen not to do them in this PR (else I'd need to actually split out and clobber the components, which was beyond what I wanted to do in this PR).

File: src/Notifier.js (hunk 1 above)
Comment (Kegsay) [inline src/Notifier.js]: I agree other apps will want to present their notifications differently. Exposing this as a React-style component though is incredibly misleading. I'm completely happy to expose an API on the `Notifier` to allow other apps to configure the look and feel as they wish.

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (dbkr) [inline src/components/structures/MatrixChat.js]: Oh I see, this makes more sense.

File: src/Notifier.js (hunk 1 above)
Comment (dbkr) [inline src/Notifier.js]: The plan was to move it over to be a Module although this PR removes the Module system. I'm somewhat concerned that we're dealing with tradeoff's we've made in the name of making a flexible, re-usable SDK by removing all the flexibility and worried that we're going to end up in the same situation again when we add the flexibility back in again.

Comment (Kegsay) [inline src/Notifier.js]: The flexibility is provided through the component system, nothing more. We allow people to either clobber specific components or write their own, and we can design reusable components which allow them to use the same code that we use in Vector. The exposed API surface is simple and understandable to an outsider (it's a dictionary of `string => Component`, and is accessed via `sdk.getComponent("string")`). I like this simplicity. I don't think we need to add anything to this.

I think it is a mistake to add more layers of complexity by adding more systems to make "flexible" logic classes (e.g. A React-classless `Notifier` which looks and smells like a React class but isn't one, a `Modulator` system which is used precisely once). Logic classes are a lot easier to dependency inject (e.g. through constructor injection) than a tree of UI components, which is why `getComponent()` makes a lot of sense for UI components but makes less sense for logic classes IMO.

**Pull request merged**